include/mqt-core/dd/Package.hpp

```diff
@@ -1985,24 +2026,51 @@ template <class Config> class Package {
       return CachedEdge<Node>::zero();
     }
 
-    if (std::none_of(eliminate.begin(), eliminate.end(),
+    // If `a` is the identity matrix or there is nothing left to eliminate,
+    // then simply return `a`
+    if (a.isIdentity() ||
+        std::none_of(eliminate.begin(),
+                     eliminate.begin() +
+                         static_cast<std::vector<bool>::difference_type>(level),
                      [](bool v) { return v; })) {
       return CachedEdge<Node>{a.p, aWeight};
     }
 
-    if (a.isIdentity()) {
-      const auto elims =
-          std::count(eliminate.begin(),
-                     eliminate.begin() + static_cast<int64_t>(level), true);
-      return CachedEdge<Node>{a.p, aWeight * std::pow(2, elims)};
-    }
-
     const auto v = a.p->v;
     if (eliminate[v]) {
+      // Lookup nodes marked for elimination in the compute table if all
+      // lower level qubits are eliminated as well: if the trace has already
+      // been computed, return the result
+      auto& computeTable = getTraceComputeTable<Node>();
+      if (std::all_of(
+              eliminate.begin(),
+              eliminate.begin() +
+                  static_cast<std::vector<bool>::difference_type>(level),
+              [](bool e) { return e; })) {
```

Comment:
Just a quick side comment (besides the currently failing unittest): I think you can extract this computation to avoid repeating it at the end of this block.
```c++
const auto computeFullTrace = std::all_of(
              eliminate.begin(),
              eliminate.begin() +
                  static_cast<std::vector<bool>::difference_type>(level),
              [](bool e) { return e; });
```
or something similar.

Reply:
Done, thanks!